### PR TITLE
Basic QiDao Connectors

### PR DESCRIPTION
**Pull request by royalaid**
This PR adds connectors that allow the creation of deposits, withdrawals from, borrowing against, and repayment to QiDao vaults.
- [x] Review and approve connector. @pradyuman-verma 
- [ ] Deploy the contracts and comment the addresses.
- [ ] Created resolver - [here](https://github.com/Instadapp/dsa-resolvers/pull/45)

### CODE REVIEW

**Comment by pradyuman-verma** [comment]
Hey @royalaid, Can you rebase this branch with our latest commits?
As the repo has enabled Typescript support and many other changes, it will be easy for us to review it then!

File: scripts/tests/addLiquidity.ts [hunk 1]
```diff
@@ -39,40 +40,53 @@ const tokenMapping: Record<string, any> = {
...
+      process: async function (owner: Signer | Provider, to: any, amt: any) {
+        const contract = new ethers.Contract(this.address, this.abi, owner);
+        await mineTx(contract.transfer(to, amt));
+      },
+    }
+  }
 };
 
-export async function addLiquidity(tokenName: string, address: any, amt: any) {
+export async function addLiquidity(tokenName: string, address: any, amt: any, chain: string) {
```
Comment (pradyuman-verma) [inline scripts/tests/addLiquidity.ts]: No need to pass `chain` as a params here!

File: scripts/tests/addLiquidity.ts [hunk 2]
```diff
@@ -39,40 +40,53 @@ const tokenMapping: Record<string, any> = {
...
   const [signer] = await ethers.getSigners();
+  const _chain = chain ? chain : 'eth';
```
Comment (pradyuman-verma) [inline scripts/tests/addLiquidity.ts]:
```suggestion
  const _chain = String(process.env.networkType);
```

File: test/polygon/qidao/qidao.test.ts [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+      await wallet0.sendTransaction({
+        to: dsaWallet0.address,
+        value: parseEther("10"),
+      });
+      expect(await ethers.provider.getBalance(dsaWallet0.address)).to.be.gte(
+        parseEther("10")
+      );
+    });
+    it("Deposit LINK into DSA wallet", async function() {
+      await addLiquidity("link", dsaWallet0.address, parseEther("100"), "polygon")
```
Comment (pradyuman-verma) [inline test/polygon/qidao/qidao.test.ts]:
```suggestion
      await addLiquidity("link", dsaWallet0.address, parseEther("100"))
```

File: scripts/tests/addLiquidity.ts [hunk 4]
```diff
@@ -8,14 +8,15 @@ const mineTx = async (tx: any) => {
...
 const tokenMapping: Record<string, any> = {
-  usdc: {
-    impersonateSigner: "0xfcb19e6a322b27c06842a71e8c725399f049ae3a",
-    address: "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
-    abi: [
-      "function mint(address _to, uint256 _amount) external returns (bool);",
-    ],
-    process: async function(owner: Signer | Provider, to: any, amt: any) {
-      const contract = new ethers.Contract(this.address, this.abi, owner);
+  eth:{
```
Comment (pradyuman-verma) [inline scripts/tests/addLiquidity.ts]:
```suggestion
  mainnet:{
```

File: test/polygon/qidao/qidao.test.ts [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+  let connector: any;
+  let masterSigner: Signer;
+
+  before(async () => {
+    await hre.network.provider.request({
+      method: "hardhat_reset",
+      params: [
+        {
+          forking: {
+            jsonRpcUrl: hre.config.networks.hardhat.forking.url,
```
Comment (pradyuman-verma) [inline test/polygon/qidao/qidao.test.ts]:
```suggestion
          forking: {
          // @ts-ignore
            jsonRpcUrl: hre.config.networks.hardhat.forking.url,
```

File: test/polygon/qidao/qidao.test.ts [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+const { ethers } = hre;
+
+describe("QiDao", function() {
+  const connectorName = "QIDAO-TEST-A";
+
+  let wallet0: any, wallet1: any;
+  let dsaWallet0: any;
+  let instaConnectorsV2: Contract;
+  let connector: any;
+  let masterSigner: Signer;
```
Comment (pradyuman-verma) [inline test/polygon/qidao/qidao.test.ts]:
```suggestion
  let wallet0: Signer, wallet1: Signer;
  let dsaWallet0: Contract;
  let instaConnectorsV2: Contract;
  let connector: Contract;
  let masterSigner: Signer;
```

File: scripts/constant/qidao/vaults.ts [hunk 7]
```diff
@@ -0,0 +1,8 @@
+export const vaults = {
+  matic: {
+    address: "0xa3fa99a148fa48d14ed51d610c367c61876997f1"
+  },
+  link: {
+    address: "0x61167073E31b1DAd85a3E531211c7B8F1E5cAE72"
+  }
+}
```
Comment (pradyuman-verma) [inline scripts/constant/qidao/vaults.ts]: Can we move this into `root/tests/polygon/qidao/`

File: test/polygon/qidao/qidao.test.ts [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+
+const { expect } = require("chai");
+const hre = require("hardhat");
+const { abis } = require("../../../scripts/constant/abis");
+const { addresses } = require("../../../scripts/tests/polygon/addresses");
+const { tokens }  = require("../../../scripts/tests/polygon/tokens");
+const { deployAndEnableConnector } = require("../../../scripts/tests/deployAndEnableConnector");
+const { getMasterSigner } = require("../../../scripts/tests/getMasterSigner");
+const { buildDSAv2 } = require("../../../scripts/tests/buildDSAv2");
+const ConnectV2QiDaoPolygon = require("../../../artifacts/contracts/polygon/connectors/qidao/main.sol/ConnectV2QiDaoPolygon.json");
```
Comment (pradyuman-verma) [inline test/polygon/qidao/qidao.test.ts]:
```suggestion
const {ConnectV2QiDaoPolygon__factory} from "../../../typechain"
```

File: test/polygon/qidao/qidao.test.ts [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+    });
+    [wallet0, wallet1] = await ethers.getSigners();
+    masterSigner = await getMasterSigner();
+    instaConnectorsV2 = await ethers.getContractAt(
+      abis.core.connectorsV2,
+      addresses.core.connectorsV2
+    );
+    connector = await deployAndEnableConnector({
+      connectorName,
+      contractArtifact: ConnectV2QiDaoPolygon,
```
Comment (pradyuman-verma) [inline test/polygon/qidao/qidao.test.ts]:
```suggestion
      contractArtifact: ConnectV2QiDaoPolygon__factory,
```

**Comment by royalaid** [comment]
image

File: contracts/polygon/connectors/qidao/helpers.sol [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+            finalBal = tokenContract.balanceOf(address(this));
+        }
+    }
+
+    function _borrow(
+        address vaultAddress,
+        uint256 vaultId,
+        uint256 amt
+    ) public {
+        erc20StablecoinInterface vault = erc20StablecoinInterface(vaultAddress);
```
Comment (pradyuman-verma) [inline contracts/polygon/connectors/qidao/helpers.sol:90]: Why are we creating its instance again and again? Shouldn't we make one instance and use it!

File: test/polygon/qidao/qidao.test.ts [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+      );
+    });
+    // it("Should borrow and payback half DAI from Aave V2", async function() {
+    //   const amt = parseEther("100"); // 100 DAI
+    //   // const setId = "83478237";
+    //   await addLiquidity("dai", dsaWallet0.address, parseEther("1"));
+    //   let spells = [
+    //     {
+    //       connector: connectorName,
+    //       method: "borrow",
```
Comment (pradyuman-verma) [inline test/polygon/qidao/qidao.test.ts]: Why are these commented ?

Comment (royalaid) [reply test/polygon/qidao/qidao.test.ts]: Copy and paste hold over from porting the old Aave tests. Will delete

File: contracts/polygon/connectors/qidao/main.sol [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+    {
+        uint256 _amt = getUint(getAmtId, amt);
+        uint256 _vaultId = getUint(getVaultId, vaultId);
+
+        _borrow(vaultAddress, _vaultId, _amt);
+
+        setUint(setAmtId, _amt);
+        setUint(getVaultId, _vaultId);
+
+        _eventName = "LogBorrow(uint256, uint256, uint256, uint256, uint256, uint256);";
```
Comment (pradyuman-verma) [inline contracts/polygon/connectors/qidao/main.sol]: We cann't have spaces here.
```suggestion
        _eventName = "LogBorrow(uint256,uint256,uint256,uint256,uint256,uint256);";
```
```suggestion
        _eventName = "LogBorrow(uint256, uint256, uint256, uint256, uint256, uint256);";
```

File: contracts/polygon/connectors/qidao/main.sol [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+        uint256 _amt = getUint(getAmtId, amt);
+        uint256 _vaultId = getUint(getVaultId, vaultId);
+
+        (uint256 initialBal, uint256 finalBal) = _withdraw(token, vaultAddress, _vaultId, _amt);
+
+        _amt = sub(finalBal, initialBal);
+
+        setUint(setAmtId, _amt);
+
+        _eventName = "LogWithdrawCollateral(uint256, uint256, uint256, uint256, uint256, uint256)";
```
Comment (pradyuman-verma) [inline contracts/polygon/connectors/qidao/main.sol]: No spaces here too

File: contracts/polygon/connectors/qidao/main.sol [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+    {
+        uint256 _amt = getUint(getAmtId, amt);
+        uint256 _vaultId = getUint(getVaultId, vaultId);
+
+        _deposit(token, vaultAddress, _vaultId, _amt);
+
+        setUint(setAmtId, _amt);
+        setUint(getVaultId, _vaultId);
+
+        _eventName = "LogDepositCollateral(uint256, uint256, uint256, uint256, uint256, uint256)";
```
Comment (pradyuman-verma) [inline contracts/polygon/connectors/qidao/main.sol]: no spaces

File: contracts/polygon/connectors/qidao/main.sol [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+        address vaultAddress,
+        uint256 vaultId,
+        uint256 getId
+    )
+        external
+        payable
+        returns (string memory _eventName, bytes memory _eventParam)
+    {
+        uint256 _vaultId = _destroyVault(vaultAddress, vaultId, getId);
+        _eventName = "LogDestroyVault(uint256, address)";
```
Comment (pradyuman-verma) [inline contracts/polygon/connectors/qidao/main.sol]: no spaces

File: contracts/polygon/connectors/qidao/main.sol [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+     * @param vaultAddress The address of the vault smart contract for the token/asset
+     * @param setId ID of the created vault.
+     */
+    function createVault(address vaultAddress, uint256 setId)
+        external
+        payable
+        returns (string memory _eventName, bytes memory _eventParam)
+    {
+        uint256 _vaultId = getVaultId(vaultAddress, setId);
+        _eventName = "LogCreateVault(uint256, address)";
```
Comment (pradyuman-verma) [inline contracts/polygon/connectors/qidao/main.sol]: no spaces

File: contracts/polygon/connectors/qidao/main.sol [hunk 17]
```diff
@@ -0,0 +1,232 @@
...
+    {
+        uint256 _amt = getUint(getAmtId, amt);
+        uint256 _vaultId = getUint(getVaultId, vaultId);
+
+        _payback(vaultAddress, _vaultId, _amt);
+
+        setUint(setAmtId, _amt);
+        setUint(getVaultId, _vaultId);
+
+        _eventName = "LogPayBack(uint256, uint256, uint256, uint256, uint256, uint256)";
```
Comment (pradyuman-verma) [inline contracts/polygon/connectors/qidao/main.sol]: no spaces

File: contracts/polygon/connectors/qidao/helpers.sol [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+
+    function _deposit(
+        address token,
+        address vaultAddress,
+        uint256 vaultId,
+        uint256 amt
+    ) public {
+        bool isEth = token == maticAddr;
+
+        if (isEth) {
```
Comment (pradyuman-verma) [inline contracts/polygon/connectors/qidao/helpers.sol]:
```suggestion
    ) public {
        bool isMatic = token == maticAddr;

        if (isMatic) {
```

File: contracts/polygon/connectors/qidao/helpers.sol [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+
+    function _withdraw(
+        address token,
+        address vaultAddress,
+        uint256 vaultId,
+        uint256 amt
+    ) public returns (uint256 initialBal, uint256 finalBal) {
+        bool isEth = token == maticAddr;
+
+        if (isEth) {
```
Comment (pradyuman-verma) [inline contracts/polygon/connectors/qidao/helpers.sol]:
```suggestion
        bool isMatic = token == maticAddr;

        if (isMatic) {
```

File: test/polygon/qidao/qidao.test.ts (hunk 11 above)
Comment (pradyuman-verma) [reply test/polygon/qidao/qidao.test.ts]: Okay

File: contracts/polygon/connectors/qidao/main.sol [hunk 20]
```diff
@@ -0,0 +1,232 @@
...
+     * @param getAmtId ID to retrieve amt.
+     * @param setAmtId ID stores the amount of tokens withdrawn.
+     */
+    function deposit(
+        address token,
+        address vaultAddress,
+        uint256 vaultId,
+        uint256 amt,
+        uint256 getVaultId,
+        uint256 setVaultId,
```
Comment (pradyuman-verma) [inline contracts/polygon/connectors/qidao/main.sol]: Unused param `setVaultId`

File: contracts/polygon/connectors/qidao/main.sol [hunk 21]
```diff
@@ -0,0 +1,232 @@
...
+     * @param getAmtId ID to retrieve amt.
+     * @param setAmtId ID stores the amount of tokens withdrawn.
+     */
+    function withdraw(
+        address token,
+        address vaultAddress,
+        uint256 vaultId,
+        uint256 amt,
+        uint256 getVaultId,
+        uint256 setVaultId,
```
Comment (pradyuman-verma) [inline contracts/polygon/connectors/qidao/main.sol]: Unused param `setVaultId`

File: contracts/polygon/connectors/qidao/main.sol [hunk 22]
```diff
@@ -0,0 +1,232 @@
...
+     * @param setVaultId ID stores the vault being interacted with.
+     * @param getAmtId ID to retrieve amt.
+     * @param setAmtId ID stores the amount of tokens withdrawn.
+     */
+    function borrow(
+        address vaultAddress,
+        uint256 vaultId,
+        uint256 amt,
+        uint256 getVaultId,
+        uint256 setVaultId,
```
Comment (pradyuman-verma) [inline contracts/polygon/connectors/qidao/main.sol]: Unused param `setVaultId`

File: contracts/polygon/connectors/qidao/main.sol [hunk 23]
```diff
@@ -0,0 +1,232 @@
...
+     * @param setVaultId ID stores the vault being interacted with.
+     * @param getAmtId ID to retrieve amt.
+     * @param setAmtId ID stores the amount of tokens withdrawn.
+     */
+    function payback(
+        address vaultAddress,
+        uint256 vaultId,
+        uint256 amt,
+        uint256 getVaultId,
+        uint256 setVaultId,
```
Comment (pradyuman-verma) [inline contracts/polygon/connectors/qidao/main.sol]: Unused param `setVaultId`

File: contracts/polygon/connectors/qidao/main.sol [hunk 24]
```diff
@@ -0,0 +1,226 @@
...
+
+        setUint(setAmtId, _amt);
+        setUint(getVaultId, _vaultId);
+
+        _eventName = "LogDepositCollateral(uint256,uint256,uint256,uint256,uint256)";
+        _eventParam = abi.encode(
+            _vaultId,
+            _amt,
+            getVaultId,
+            setVaultId,
```
Comment (pradyuman-verma) [inline contracts/polygon/connectors/qidao/main.sol]: Forgot to remove this!
```suggestion
          
```

File: contracts/polygon/connectors/qidao/main.sol [hunk 25]
```diff
@@ -0,0 +1,226 @@
...
+
+        setUint(setAmtId, _amt);
+        setUint(getVaultId, _vaultId);
+
+        _eventName = "LogPayBack(uint256,uint256,uint256,uint256,uint256)";
+        _eventParam = abi.encode(
+            _vaultId,
+            _amt,
+            getVaultId,
+            setVaultId,
```
Comment (pradyuman-verma) [inline contracts/polygon/connectors/qidao/main.sol]: this too
```suggestion

```

File: test/polygon/qidao/qidao.test.ts [hunk 26]
```diff
@@ -0,0 +1,168 @@
...
+      const spells = [
+        {
+          connector: connectorName,
+          method: "createVault",
+          args: [vaults.matic.address, setVaultId],
+        },
+        {
+          connector: connectorName,
+          method: "deposit",
+          args: [tokens.matic.address, vaults.matic.address, 0, amt, setVaultId, 0, 0, 0],
```
Comment (pradyuman-verma) [inline test/polygon/qidao/qidao.test.ts]: args count should be seven right!
as we've removed `setVaultId`.
```suggestion
          args: [tokens.matic.address, vaults.matic.address, 0, amt, 0, 0, 0],
```

File: test/polygon/qidao/qidao.test.ts [hunk 27]
```diff
@@ -0,0 +1,168 @@
...
+        },
+        {
+          connector: connectorName,
+          method: "borrow",
+          args: [vaults.matic.address, 0, brwAmt, setVaultId, 0, 0 , 0]
```
Comment (pradyuman-verma) [inline test/polygon/qidao/qidao.test.ts]: here same

File: test/polygon/qidao/qidao.test.ts [hunk 28]
```diff
@@ -0,0 +1,168 @@
...
+        },
+        {
+          connector: connectorName,
+          method: "payback",
+          args: [vaults.matic.address, 0, brwAmt, setVaultId, 0, 0 , 0],
```
Comment (pradyuman-verma) [inline test/polygon/qidao/qidao.test.ts]: here!

File: test/polygon/qidao/qidao.test.ts [hunk 29]
```diff
@@ -0,0 +1,168 @@
...
+        },
+        {
+          connector: connectorName,
+          method: "withdraw",
+          args: [tokens.matic.address, vaults.matic.address, 0, amt.mul(995).div(1000), setVaultId, 0, 0, 0],
```
Comment (pradyuman-verma) [inline test/polygon/qidao/qidao.test.ts]: here too!

**Comment by pradyuman-verma** [comment]
@royalaid  Can you push a dummy commit? We have changed the checks, and I think now they will pass.

**Pull request closed without merging**